ARTEMIS-2226 last consumer connection should close the previous consu…

onlyMIT opened the pull request:
If use MQTT protocol, multiple consumers using the same clientId in the cluster, the last consumer connection should close the previous consumer connection! 
There are two solutions to this issue:

> 1. Add a new cluster notification when the client’s connections connect to Artemis, close existing connections that use the same clientId when processing cluster-notification. This solution is the best way to address the MQTT agreement.But I don't think it's necessary to solve this problem by sending a cluster-notification every time client connect.

> 2.The solution is this RP used, close existing connections that use the same clientId when processing the CONSUMER_CREATE  cluster-notification. No additional cluster-notification added. Only when a consumer using the same clientId consumes the same topic will it trigger a close connection.

franz1981 commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/postoffice/impl/PostOfficeImpl.java:
```diff
@@ -343,6 +347,18 @@ public void onNotification(final Notification notification) {
                      Queue queue = (Queue) binding.getBindable();
+                     //If the client ID represents a client already connected to the server then the server MUST disconnect the existing client.
+                     //Avoid consumers with the same client ID in the cluster appearing at different nodes at the same time
+                     SimpleString protocolName = props.getSimpleStringProperty(ManagementHelper.HDR_PROTOCOL_Name);
```
I'm not that familiar with the MQTT protocol, but I can give you an idea: instead of using a SimpleString property, that is costly to be compared/transferred/encoded/decoded why not define a single byte (or short to be more future-proof) property? 
It would be less human-readable, but more cheaper for a machine: in addition, a Short/Byte tends to be pooled on the ranges of -127<->128 AFAIK or something similar, so only the property name + HashMap entry will produce some garbage.

onlyMIT replied:
Thanks, very nice advice. i will done

michaelandrepearce commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/postoffice/impl/PostOfficeImpl.java:
```diff
@@ -343,6 +347,18 @@ public void onNotification(final Notification notification) {
+                     //Avoid consumers with the same client ID in the cluster appearing at different nodes at the same time
+                     SimpleString protocolName = props.getSimpleStringProperty(ManagementHelper.HDR_PROTOCOL_Name);
+                     if (protocolName != null && protocolName.equals(ProtocolName.MQTT_PROTOCOL_NAME.getProtocolName())) {
+                        String clientId = props.getSimpleStringProperty(ManagementHelper.HDR_CLIENT_ID).toString();
```
This is protocol specific logic. It is not suitable to have this in the core modules.Any mqtt logic MUST be in the protocol module only.

michaelandrepearce commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/ServerSessionImpl.java:
```diff
@@ -540,6 +541,12 @@ public ServerConsumer createConsumer(final long consumerID,
          }
 
+         String protocolName = remotingConnection.getProtocolName();
+         if (protocolName != null && protocolName.equals(ProtocolName.MQTT_PROTOCOL_NAME.getProtocolName().toString())) {
```
This is protocol specific logic. This MUST not ever live in core modules. Protocol specific logic can ONLY reside in its protocol module

michaelandrepearce commented on artemis-core-client/src/main/java/org/apache/activemq/artemis/api/core/management/ManagementHelper.java:
```diff
@@ -84,6 +84,10 @@
 
    public static final SimpleString HDR_MESSAGE_ID = new SimpleString("_AMQ_Message_ID");
 
+   public static final SimpleString HDR_PROTOCOL_Name = new SimpleString("_AMQ_ProtocolName");
```
static final field should be all upper case. HDR_PROTOCOL_NAME
key should be '_' separated _AMQ_Protocol_Name

michaelandrepearce commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/protocol/ProtocolName.java:
```diff
@@ -0,0 +1,38 @@
+
+public enum ProtocolName {
+
+   MQTT_PROTOCOL_NAME(SimpleString.toSimpleString("MQTT")),
```
-1, Core should not know about specific protocols.

onlyMIT replied:
Got it thanks.  Since the processing of cluster-notification is in the Core, and the Core does not depend on the specific protocol code. I will think about how to adjust the code structure and then submit a version. Maybe need a listener or a callback

onlyMIT commented:
@michaelandrepearce @franz1981  This PR has updated.  
Because of core modules  should not know about specific protocols,  can't map protocol to byte in CORE module，so protocol name still use SimpleString property in cluster-notification

michaelandrepearce commented on artemis-protocols/artemis-mqtt-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/mqtt/MQTTProtocolManager.java:
```diff
@@ -62,11 +73,44 @@
+            //Avoid consumers with the same client ID in the cluster appearing at different nodes at the same time
+            Collection<Consumer> consumersSet = queue.getConsumers();
+            for (Consumer consumer : consumersSet) {
+               ServerConsumerImpl serverConsumer = (ServerConsumerImpl) consumer;
```
Cast to interface not concrete class.

michaelandrepearce replied:
Also should check instanceof, as by contract, it may not be a ServerConsumer

onlyMIT replied:
I also disagree with putting this method into the `ServerConsumer` interface. Using instanceof is a good suggestion.

michaelandrepearce replied:
what method? if youre talking re remoting connection you shouldnt be even have added that new method, you simply want to close previous consumers, you therefor should be calling close method on the consumer.

michaelandrepearce commented on artemis-protocols/artemis-mqtt-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/mqtt/MQTTProtocolManager.java:
```diff
@@ -62,11 +73,44 @@
+            for (Consumer consumer : consumersSet) {
+               ServerConsumerImpl serverConsumer = (ServerConsumerImpl) consumer;
+               if (clientId.equals(serverConsumer.getConnectionClientID()))
+                  serverConsumer.getRemotingConnection().destroy();
```
The new getRemotingConnection isn't needed.
 You should simply be closing the consumer, this will then mean all bits around closing a consumer are safely done, and also other notifications sent.

onlyMIT replied:
Our goal is to avoid simultaneous occurrence of multiple MQTT connections with the same clientId, not to prevent MQTT consumers with the same clientId to subscribe the same topic.

described in the [two solutions](https://github.com/apache/activemq-artemis/pull/2528#issue-248449219) .  Adding a new cluster notification when the client’s connections connect to Artemis,
it is the best way to address the MQTT agreement, but the cost is too great. So it put the logic in CREATE_CONSUMER cluster-notification.

The following code description, before each processing of the mqtt connection request in artemis, check whether there is a connection with the same clientId in the node, and if so, close it.
Similarly, in the cluster, we also want to avoid the connection of the same clientId. Executing this logic only in CREATE_CONSUMER is just that I think it can reduce resource consumption and avoid problems, although this is not the way to match the MQTT specification.
```
private String validateClientId(String clientId, boolean cleanSession) {
      if (clientId == null || clientId.isEmpty()) {
         // [MQTT-3.1.3-7] [MQTT-3.1.3-6] If client does not specify a client ID and clean session is set to 1 create it.
         if (cleanSession) {
            clientId = UUID.randomUUID().toString();
         } else {
            // [MQTT-3.1.3-8] Return ID rejected and disconnect if clean session = false and client id is null
            return null;
         }
      } else {
         MQTTConnection connection = session.getProtocolManager().addConnectedClient(clientId, session.getConnection());

         if (connection != null) {
            // [MQTT-3.1.4-2] If the client ID represents a client already connected to the server then the server MUST disconnect the existing client
            connection.disconnect(false);
         }
      }
      return clientId;
   }
```

michaelandrepearce replied:
But your current logic is simply find all consumers with matching clientId and kill beneath it, their connection. You may as well simply close down the consumer

michaelandrepearce replied:
Why is:

Adding a new cluster notification when the client’s connections connect to Artemis,
it is the best way to address the MQTT agreement, but the cost is too great

Whats the cost? To me if you want the behaviour on connection behaviour this truely would be better, rather than working on consumer

michaelandrepearce replied:
we already have a connection notification, just need to add the generic extra data as you were going to on the consumer

   private void sendConnectionNotification(final RemotingConnection connection, final CoreNotificationType type) {
      final ManagementService managementService = getManagementService();

      if (managementService != null && sendConnectionNotifications) {
         try {
            final TypedProperties props = new TypedProperties();
            props.putSimpleStringProperty(ManagementHelper.HDR_CONNECTION_NAME, SimpleString.toSimpleString(connection.getID().toString()));
            props.putSimpleStringProperty(ManagementHelper.HDR_REMOTE_ADDRESS, SimpleString.toSimpleString(connection.getRemoteAddress()));
            props.putSimpleStringProperty(ManagementHelper.HDR_CLIENT_ID, SimpleString.toSimpleString(connection.getClientID()));
            props.putSimpleStringProperty(ManagementHelper.HDR_PROTOCOL_NAME, SimpleString.toSimpleString(connection.getProtocolName()));

            managementService.sendNotification(new Notification(null, type, props));
         } catch (Exception e) {
            logger.warn("Error sending notification: " + type, e.getMessage(), e);
         }
      }
   }

and then act just at the connection level.

onlyMIT replied:
This code seems to be easily misunderstood, I am trying to find out if there is a better way to handle this logic.

onlyMIT replied:
Thanks for your answer，i not find the connection notification，if we had we can use it.

michaelandrepearce replied:
@onlyMIT but re your top level options:

> Add a new cluster notification when the client’s connections connect to Artemis, close existing connections that use the same clientId when processing cluster-notification. This solution is the best way to address the MQTT agreement.But I don't think it's necessary to solve this problem by sending a cluster-notification every time client connect.

There already is a cluster-notification sent every time a client connects if using NotificationActiveMQServerPlugin, so if anything simply we could just make that more built in, like consumer one is, and simply make it enable the  flag sendConnectionNotifications, if MQTT protocol is used (keeping it as is today disabling for those who have not enabled MQTT)

onlyMIT replied:
got it thanks

michaelandrepearce replied:
Looking some more when an connection with MQTT is created the MQTTProtocolHandler is called, so actually may not even need notification could just hook into that logic and flow

michaelandrepearce replied:
e.g.

from
```
@Override
   public ConnectionEntry createConnectionEntry(Acceptor acceptorUsed, Connection connection) {
      try {
         MQTTConnection mqttConnection = new MQTTConnection(connection);
         ConnectionEntry entry = new ConnectionEntry(mqttConnection, null, System.currentTimeMillis(), MQTTUtil.DEFAULT_KEEP_ALIVE_FREQUENCY);

         NettyServerConnection nettyConnection = ((NettyServerConnection) connection);
         MQTTProtocolHandler protocolHandler = nettyConnection.getChannel().pipeline().get(MQTTProtocolHandler.class);
         protocolHandler.setConnection(mqttConnection, entry);

         return entry;
      } catch (Exception e) {
         log.error(e);
         return null;
      }
   }
```

to something like: (replacing the code added with the logic you think is needed)
```
@Override
   public ConnectionEntry createConnectionEntry(Acceptor acceptorUsed, Connection connection) {
      try {
         MQTTConnection mqttConnection = new MQTTConnection(connection);
         ConnectionEntry entry = new ConnectionEntry(mqttConnection, null, System.currentTimeMillis(), MQTTUtil.DEFAULT_KEEP_ALIVE_FREQUENCY);

         NettyServerConnection nettyConnection = ((NettyServerConnection) connection);
         MQTTProtocolHandler protocolHandler = nettyConnection.getChannel().pipeline().get(MQTTProtocolHandler.class);
         protocolHandler.setConnection(mqttConnection, entry);

         String clientID = connection.getProtocolConnection().getClientID();
         MQTTConnection existingMQTTConnection = connectedClients.get(clientID);
         if (existingMQTTConnection != null) {
            existingMQTTConnection.getTransportConnection().getProtocolConnection().destroy();
         }
         return entry;
      } catch (Exception e) {
         log.error(e);
         return null;
      }
   }
```

onlyMIT replied:
My goal is to close connections that have the same clientId in different nodes.The following code already done the logic which your mentioned . 
```
private String validateClientId(String clientId, boolean cleanSession) {
      if (clientId == null || clientId.isEmpty()) {
         // [MQTT-3.1.3-7] [MQTT-3.1.3-6] If client does not specify a client ID and clean session is set to 1 create it.
         if (cleanSession) {
            clientId = UUID.randomUUID().toString();
         } else {
            // [MQTT-3.1.3-8] Return ID rejected and disconnect if clean session = false and client id is null
            return null;
         }
      } else {
         MQTTConnection connection = session.getProtocolManager().addConnectedClient(clientId, session.getConnection());

         if (connection != null) {
            // [MQTT-3.1.4-2] If the client ID represents a client already connected to the server then the server MUST disconnect the existing client
            connection.disconnect(false);
         }
      }
      return clientId;
   }
```
I think I am almost revised, I will update after the test is completed.

onlyMIT replied:
And clientId is null is this place:
```
 @Override
   public ConnectionEntry createConnectionEntry(Acceptor acceptorUsed, Connection connection) {
      try {
         //New code, to kill of existingConnection if same client id.
         String clientID = connection.getProtocolConnection().getClientID();
         MQTTConnection existingMQTTConnection = connectedClients.get(clientID); 
```

clientId is set after the `createConnectionEntry `method is called.  In `MQTTConnectionManager.connect method`
```
   synchronized void connect(String cId,
                             String username,
                             byte[] passwordInBytes,
                             boolean will,
                             byte[] willMessage,
                             String willTopic,
                             boolean willRetain,
                             int willQosLevel,
                             boolean cleanSession) throws Exception {
      String clientId = validateClientId(cId, cleanSession);
      if (clientId == null) {
         session.getProtocolHandler().sendConnack(MqttConnectReturnCode.CONNECTION_REFUSED_IDENTIFIER_REJECTED);
         session.getProtocolHandler().disconnect(true);
         return;
      }

      String password = passwordInBytes == null ? null : new String(passwordInBytes, CharsetUtil.UTF_8);
      session.getConnection().setClientID(clientId);
```

michaelandrepearce commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/ServerConsumerImpl.java:
```diff
@@ -1544,4 +1545,8 @@ public String getConnectionLocalAddress() {
       return this.session.getRemotingConnection().getTransportConnection().getRemoteAddress();
    }
+
+   public RemotingConnection getRemotingConnection() {
```
Shouldnt be needed, you should just be acting on the consumer, closing the consumer via its close methods if needed.

michaelandrepearce commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/ServerSessionImpl.java:
```diff
@@ -540,6 +540,14 @@ public ServerConsumer createConsumer(final long consumerID,
+            props.putSimpleStringProperty(ManagementHelper.HDR_PROTOCOL_NAME, SimpleString.toSimpleString(protocolName));
+
+         String clientId = remotingConnection.getClientID();
+         if (clientId != null)
```
curlys on the if please

michaelandrepearce commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/ServerSessionImpl.java:
```diff
@@ -540,6 +540,14 @@ public ServerConsumer createConsumer(final long consumerID,
          }
 
+         String protocolName = remotingConnection.getProtocolName();
+         if (protocolName != null)
```
curlys on the if please

michaelandrepearce commented on artemis-protocols/artemis-mqtt-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/mqtt/MQTTProtocolManager.java:
```diff
@@ -62,11 +73,44 @@
+            String clientId = props.getSimpleStringProperty(ManagementHelper.HDR_CLIENT_ID).toString();
+            //If the client ID represents a client already connected to the server then the server MUST disconnect the existing client.
+            //Avoid consumers with the same client ID in the cluster appearing at different nodes at the same time
+            Collection<Consumer> consumersSet = queue.getConsumers();
```
Might be worth adding a new method on queue that takes an additional predicate parameter, this way can pass in a predicate of 
clientId.equals(serverConsumer.getConnectionClientID()
and return only the consumers matching, this will avoid expensive copying of all consumers if the queue has high consumer count, and repetitive iterating over those.

michaelandrepearce replied:
e.g. adding something like

```
   @Override
   public Set<Consumer> getConsumers(Predicate<Consumer> consumerPredicate) {
      Set<Consumer> consumersSet = new HashSet<>();
      for (ConsumerHolder<? extends Consumer> consumerHolder : consumers) {
         if (consumerPredicate.test(consumerHolder.consumer)) {
            consumersSet.add(consumerHolder.consumer);
         }
      }
      return consumersSet;
   }
```

onlyMIT replied:
got it

michaelandrepearce replied:
i sent this in pr to your branch

michaelandrepearce commented on artemis-protocols/artemis-mqtt-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/mqtt/MQTTSession.java:
```diff
@@ -30,10 +30,12 @@
 public class MQTTSession {
 
-   static Map<String, MQTTSessionState> SESSIONS = new ConcurrentHashMap<>();
+   private static Map<String, MQTTSessionState> SESSIONS = new ConcurrentHashMap<>();
```
make it final too?

michaelandrepearce commented on artemis-protocols/artemis-mqtt-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/mqtt/MQTTSession.java:
```diff
@@ -30,10 +30,12 @@
 public class MQTTSession {
 
-   static Map<String, MQTTSessionState> SESSIONS = new ConcurrentHashMap<>();
+   private static Map<String, MQTTSessionState> SESSIONS = new ConcurrentHashMap<>();
```
Looking at this im bit worried that this is a static in first place, e.g. its possible to have multiple embedded brokers, in that case each broker shouldnt share any state, by being static means they would.

michaelandrepearce replied:
i sent a possible solution to address this, in a pr to your branch.

michaelandrepearce commented:
@onlyMIT to help with addressing the comments i sent a pr to your branch with a number of proposed changes

onlyMIT commented:
@michaelandrepearce thanks to your review, I will address  the comments.

onlyMIT commented:
@michaelandrepearce  This PR has updated.  `SESSIONS` is still in MQTTSession,  do not move to  `MQTTProtocolHandler ` . Because SESSIONS should be shared by multiple Acceptors. 
```
public class MQTTSession {

    static Map<String, MQTTSessionState> SESSIONS = new ConcurrentHashMap<>();
   private static Map<String, MQTTSessionState> SESSIONS = new ConcurrentHashMap<>();
```

michaelandrepearce commented:
There is the factory if thats the case where a map could be passed into each handler during favtory crreate. What we cant have though is statics for this tho

michaelandrepearce commented on artemis-core-client/src/main/java/org/apache/activemq/artemis/core/server/management/NotificationService.java:
```diff
@@ -37,4 +37,8 @@
 
    void removeNotificationListener(NotificationListener listener);
 
+   default void callNotificationListener(Notification notification) {
```
This is purpose of sendNotification, any new thing should integrate with the existing way things work.

michaelandrepearce commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/server/management/impl/ManagementServiceImpl.java:
```diff
@@ -633,6 +633,20 @@ public boolean isStarted() {
    }
 
+   @Override
+   public void callNotificationListener(Notification notification) {
```
This is repeating logic of sendNotification. Please just use that.

michaelandrepearce commented on artemis-protocols/artemis-mqtt-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/mqtt/MQTTConnectionManager.java:
```diff
@@ -163,13 +163,7 @@ synchronized void disconnect(boolean failure) {
-      }
-
-      return state;
+      return MQTTSession.getSessionState(clientId, session.getServer().getIdentity());
```
-1 Ive raised this already, simply shouldnt be having static map holding state.

onlyMIT commented:
@michaelandrepearce I have addressed the comments your sent a pr to my branch. Change the clusterConsumer queue’s filter. Because it will filter CONNECTION_CONNECTED notification.

michaelandrepearce commented on artemis-core-client/src/main/java/org/apache/activemq/artemis/spi/core/protocol/RemotingConnection.java:
```diff
@@ -245,6 +245,13 @@
+   /**
+    * Called when all connection related operations are completed.
+    */
+   default void connected() throws Exception {
```
As this introduces a notification about connections created, should ensure that this is always called, and not dependent on a protocol logic.

michaelandrepearce commented on artemis-protocols/artemis-mqtt-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/mqtt/MQTTConnection.java:
```diff
@@ -216,8 +220,10 @@ public void flush() {
-      this.connected = connected;
+   @Override
+   public void connected() throws Exception {
+      remotingService.sendConnectedNotification(this);
```
This should be responsibility of core system to sending notification so connected notification is generic.

michaelandrepearce replied:
see other comments, ive dug deeper on this and seems actually the act of mqtt connected == core server session created.

michaelandrepearce commented on artemis-core-client/src/main/java/org/apache/activemq/artemis/api/core/management/CoreNotificationType.java:
```diff
@@ -47,7 +47,8 @@
    MESSAGE_DELIVERED(28),
-   MESSAGE_EXPIRED(29);
+   MESSAGE_EXPIRED(29),
+   CONNECTION_CONNECTED(30);
```
because this is in core, this must be implemented in a generic way for ALL protocols.

michaelandrepearce replied:
Going through and running this, and running with notification plugin, i notice every time this occurs, a SESSION_CREATED is also sent. 

digging deeper and looking going through the logic of the connect in MQTTConnectionManager it seems this actually connected in the MQTTConnectionManager maps over to CORE as a server session (e.g. connect == create a serversession), which means we already get a SESSION_CREATED notification with that plugin, so is it actually we want to ensure SESSION_CREATED notificaions are not reliant on plugin, e.g. bring that in... 

populating.
session.getRemotingConnection().getClientID()

michaelandrepearce replied:
So this ties in with the fact you want to ensure a single connected session (its not actually at connection level, its actually at the session level).

onlyMIT replied:
I am here to make a unified reply to your comments.

In fact, this is only the mqtt protocol need, other protocols do not.Either way, it doesn't look too good in CORE, it just tries to make it look reasonable and more readable.
```
   /**
    * Called when all connection related operations are completed.
    */
   default void connected() throws Exception {
```
I use `connected` as default not a interface method，Because we need to send cluster-notifation when the mqtt protocol connection is initialized, But other protocols don't need this method to do anything, use a empty default methods I think is appropriate.And all protocol connections have this connection completion status, and other protocols have the possibility to use this method. As for the logic that is only needed for MQTT to add a listener or callback, I think it is over-designed.

I also find SESSION_CREATED  and CONNECTION_CREATED, but i can not use them, because we need clientId, and clientId read from client message. The clientId is read from the client's connection message. The logic for reading the clientId cannot be placed in the CORE code， when create a connection or create a session.

For using callbacks or listeners, I still prefer to use the default method connected. Unless you have a better design to let me change my mind. 
the clientId is read from the client message, it seems that the message can only be sent after the protocol connection has been processed. ( MQTT protocol neet to process the CONNECT request packet, some protocols do not have a CONNECT request packet)

michaelandrepearce replied:
re

"The clientId is read from the client's connection message. The logic for reading the clientId cannot be placed in the CORE code， when create a connection or create a session."

the clientid is set on the connection BEFORE createServerSession as such it will be available when the SESSION_CREATED notification is created, 

MQTTConnectionManager lines 78, 79
```
 session.getConnection().setClientID(clientId);
 ServerSessionImpl serverSession = createServerSession(username, password);
```
just need to populate the extra field which would be available at the time via session.getRemotingConnection().getClientID() .

onlyMIT replied:
```
if (hasBrokerSessionPlugins()) {
    callBrokerSessionPlugins(plugin -> plugin.afterCreateSession(session));
}
```
SESSION_CREATED will only be sent if BrokerSessionPlugins is turned on in the configuration file.
Also the notifications of CONNECTION_CREATE and SESSIONTION_CREATE will be filtered and will not be sent to other nodes in the cluster.You can see `ClusterConnectionBridge.setupNotificationConsumer()` method, where the cluster-connection filter is set .
This is why I want to modify the filtering rules, CONNECTION_CREATE, SESSIONTION_CREATE and other connection-related notifications, only can use the `ManagementNotificationAddress` address to route messages.

onlyMIT replied:
```
private void handleNotificationMessage(ClientMessage message, CoreNotificationType ntype) throws Exception {
         // TODO - optimised this by just passing int in header - but filter needs to be extended to support IN with

         switch (ntype) {
            case BINDING_ADDED: {
               doBindingAdded(message);

               break;
            }
            case BINDING_REMOVED: {
               doBindingRemoved(message);

               break;
            }
            case CONSUMER_CREATED: {
               doConsumerCreated(message);

               break;
            }
            case CONSUMER_CLOSED: {
               doConsumerClosed(message);

               break;
            }
            case PROPOSAL: {
               doProposalReceived(message);

               break;
            }
            case PROPOSAL_RESPONSE: {
               doProposalResponseReceived(message);
               break;
            }
            case UNPROPOSAL: {
               doUnProposalReceived(message);
               break;
            }
            default: {
               throw ActiveMQMessageBundle.BUNDLE.invalidType(ntype);
            }
         }
      }
```

Only these notifications will pass between cluster nodes

michaelandrepearce commented:
As ive already said you have two options either make mqtt to ensure plugin js enabled or simply make this notification more perm as like consumer one is.

onlyMIT commented:
@michaelandrepearce As described in my comment.Even if use a plugin, we can't use notifications in the plugin like SESSION_CREATED，it can't achieve our goal, modify the SESSION_CREATED filter? I think we need to think carefully, and if we modify their filtering rules, we can't affect the existing logic. We need to conduct comprehensive testing and evaluation.I think the plugin should be where the optional code is placed, and the logic that is must required is not suitable for the plugin.

Like the first version I submitted, the logic was placed in CONSUMER_CREATE, the readability was too poor and not exactly matched with the MQTT specification; 
This version the logic was added by adding CONNECTED_CREATED. It seems that not all protocols are used, and it is not the most Good plan.
 If you want to use SESSION_CREATED, CONNECTION_CREATED,  must force the MQTT protocol to open all plugins? and modify the filter so that the SESSION_CREATED notification can communicate between the clusters. This is the risk of making the original notification rules undefined，and need to modify their filtering rules, I am not sure if modifying the filtering rules will introduce other issues.

onlyMIT commented:
Now let's solve it step by step. First, we need to determine whether to use the existing notification or add a new notification to handle this logic. If I comment, it seems that the existing notification is not suitable.
In the second step, if we use the new notification, we need to determine how to make it more reasonable.

michaelandrepearce commented:
First of all mqtt connect = create session in core. 

I dont want to introduce anything thats protocol specific. 

Im happy with current create session being a top level supported notification as this is agnostic and will cause little extra traffic compared with create consumer which already  is.

michaelandrepearce commented:
I am off now for 48 hours (pto).

michaelandrepearce commented:
As noted i sent an unpolished pr to your branch last week, to enhance and use SESSION_CREATED

As noted connected == create session in core, so im against constructing some new terminology within core notifications, here.

onlyMIT commented:
@michaelandrepearce I was on vacation last week,  started this week.

onlyMIT commented:
@michaelandrepearce move logic to SESSION_CREATE notification

michaelandrepearce commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/management/impl/ActiveMQServerControlImpl.java:
```diff
@@ -2954,7 +2955,13 @@ public void onNotification(org.apache.activemq.artemis.core.server.management.No
+      if (type == CoreNotificationType.SESSION_CREATED) {
+         TypedProperties props = notification.getProperties();
+         //SESSION_CREATED notification should not be broadcast twice
+         if (props.getIntProperty(ManagementHelper.HDR_DISTANCE) > 0) {
```
Why would this logic specific to just SESSION_CREATED, would this not be generic to all?

michaelandrepearce replied:
Also isnt cyclic between nodes in cluster already avoided, because of filter in the cluster connection bridge where it does a HDR_DISTANCE < flow.maxHops.

onlyMIT replied:
To avoid calling the broadcast twice.The original logical SESSION_CREATED notification will not be transmitted between cluster nodes, and it is now required to transfer between cluster nodes. In order to maintain consistency with the original logic, this code to prevent the node receiving the SESSION_CREATED notification from making a broadcast call again.
Keep the original logic for other notification messages not judged

michaelandrepearce commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/server/cluster/impl/ClusterConnectionBridge.java:
```diff
@@ -351,10 +355,15 @@ private String appendIgnoresToFilter(String filterString) {
    }
 
+   private String createPermissiveManagementNotificationToFilter() {
+      StringBuilder filterBuilder = new StringBuilder(ManagementHelper.HDR_NOTIFICATION_TYPE).append(" = '")
```
Whats this trying to achieve?

onlyMIT replied:
The inter-cluster notification message needs to specify the routing address to route the message. Different from the notifications of CONSUMER_CREATE and BINDING_ADD, the routing address (the address created or consumed by itself) has been determined before the notification is sent. When sending a SESSION_CREATED notification, We can't know which address to use, so I chose to use the `ManagementManagementAddress`, which is consumed by each node in the cluster-connection.
But the notification message using the `ManagementNotificationAddress` address is filtered by the filter, so I modified the filter and allow SESSION_CREATED notification to be routed to other nodes in the cluster.

onlyMIT replied:
Although it is possible to add a dedicated address to the SESSION_CREATED notification when the cluster-connection is created, it is not necessary to do so.

michaelandrepearce commented:
@onlyMIT looking better, i have some questions, which im sure all makes sense if explained, but probably a sign a little java doc might be beneficial, and would help future understanding when this PR is closed.

onlyMIT commented:
Add java doc

michaelandrepearce commented:
@onlyMIT i would like some of others in the group to review this as well.

onlyMIT commented:
@michaelandrepearce Ok, thank you very much for your help.

michaelandrepearce commented:
@jbertram @clebertsuconic and anyone else. As no one else is commenting /reviewing, and to me this looks ok, Ill look to merge this in a few days if no further review comment from anyone.

clebertsuconic commented:
@michaelandrepearce the commit message should have been edited before merged.

no big deal.. just mentioning it.
